pkg/clients/iam/iamuser.go

```diff
@@ -0,0 +1,72 @@
+package iam
+
+import (
+	"context"
+	"sort"
+
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+
+	"github.com/crossplane/provider-aws/apis/identity/v1alpha1"
+	aws "github.com/crossplane/provider-aws/pkg/clients"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+// UserClient is the external client used for IAM User Custom Resource
+type UserClient interface {
+	GetUserRequest(*iam.GetUserInput) iam.GetUserRequest
+	CreateUserRequest(*iam.CreateUserInput) iam.CreateUserRequest
+	UpdateUserRequest(*iam.UpdateUserInput) iam.UpdateUserRequest
+	DeleteUserRequest(*iam.DeleteUserInput) iam.DeleteUserRequest
+	ListGroupsForUserRequest(*iam.ListGroupsForUserInput) iam.ListGroupsForUserRequest
+	AddUserToGroupRequest(*iam.AddUserToGroupInput) iam.AddUserToGroupRequest
+	RemoveUserFromGroupRequest(*iam.RemoveUserFromGroupInput) iam.RemoveUserFromGroupRequest
+}
+
+// NewUserClient returns a new client using AWS credentials as JSON encoded data.
+func NewUserClient(ctx context.Context, credentials []byte, region string, auth awsclients.AuthMethod) (UserClient, error) {
+	cfg, err := auth(ctx, credentials, awsclients.DefaultSection, region)
+	if cfg == nil {
+		return nil, err
+	}
+	return iam.New(*cfg), nil
+}
+
+// LateInitializeUser fills the empty fields in *v1alpha1.User with
+// the values seen in iam.User.
+func LateInitializeUser(in *v1alpha1.IAMUserParameters, user *iam.User) {
+	if user == nil {
+		return
+	}
+
+	in.Path = awsclients.LateInitializeStringPtr(in.Path, user.Path)
+	if user.PermissionsBoundary != nil {
+		in.PermissionsBoundary = awsclients.LateInitializeStringPtr(in.PermissionsBoundary, user.PermissionsBoundary.PermissionsBoundaryArn)
+	}
+
+	for _, tag := range user.Tags {
+		in.Tags = append(in.Tags, v1alpha1.Tag{Key: *tag.Key, Value: *tag.Value})
+	}
```

Comment:
I think this will constantly append the same set of tags on every reconcile. For example presuming the external resource is tagged `cool: true`, each time we observe it call `LateInitializeUser` and append `cool: true` to our slice of tags, even if that tag is already there, thus growing the slice of tags indefinitely.

I believe this bug exists in the `IAMRole` controller too: https://github.com/crossplane/provider-aws/pull/141/files#diff-f4cf5668c6e624de813a694175ff2b28R111